File: Middleware/BugTracking.js
const fs = require('fs');
const uuid = require('uuid');
const sqlite3 = require('sqlite3').verbose();
const { createLogger, transports, format } = require('winston');

class BugTracker {
    constructor(saveLocation) {
        this.saveLocation = saveLocation;
        this.bugReportsFile = `${saveLocation}/bug_reports.db`;
        this._initializeDatabase();
        this.logger = this._setupLogger();
    }

    _initializeDatabase() {
        try {
            // Initialize database and table if not exists
            const db = new sqlite3.Database(this.bugReportsFile);
            db.serialize(() => {
                db.run(`CREATE TABLE IF NOT EXISTS bugs
                        (id TEXT PRIMARY KEY, description TEXT, status TEXT)`);
            });
            db.close();
        } catch (e) {
            this.logger.error(`Error initializing database: ${e}`);
        }
    }

    _setupLogger() {
        const logger = createLogger({
            level: 'debug',
            format: format.combine(
                format.timestamp(),
                format.simple()
            ),
            transports: [
                new transports.File({ filename: `${this.saveLocation}/bug_tracker.log` })
            ]
        });
        return logger;
    }

    reportBug(bugDescription) {
        try {
            if (!bugDescription) {
                throw new Error("Bug description cannot be empty.");
            }
            // Store bug reports in the database
            const bugId = uuid.v4();
            const db = new sqlite3.Database(this.bugReportsFile);
            db.run("INSERT INTO bugs (id, description, status) VALUES (?, ?, ?)", [bugId, bugDescription, "Open"], (err) => {
                if (err) {
                    this.logger.error(`Error reporting bug: ${err}`);
                    console.log("An error occurred while reporting the bug.");
                } else {
                    this.logger.info(`Bug reported successfully. Bug ID: ${bugId}`);
                    console.log(`Bug reported successfully. Bug ID: ${bugId}`);
                }
            });
            db.close();
        } catch (e) {
            this.logger.error(`Error reporting bug: ${e}`);
            console.log("An error occurred while reporting the bug.");
        }
    }

    viewBugs() {
        try {
            // View all reported bugs
            const db = new sqlite3.Database(this.bugReportsFile);
            db.all("SELECT * FROM bugs", (err, bugs) => {
                if (err) {
                    this.logger.error(`Error viewing bugs: ${err}`);
                    console.log("An error occurred while viewing bugs.");
                } else {
                    if (bugs && bugs.length > 0) {
                        bugs.forEach(bug => {
                            console.log(`${bug.id} - ${bug.description} (${bug.status})`);
                        });
                    } else {
                        console.log("No bugs reported.");
                    }
                }
            });
            db.close();
        } catch (e) {
            this.logger.error(`Error viewing bugs: ${e}`);
            console.log("An error occurred while viewing bugs.");
        }
    }

    fixBug(bugId) {
        try {
            const db = new sqlite3.Database(this.bugReportsFile);
            db.get("SELECT * FROM bugs WHERE id=?", [bugId], (err, bug) => {
                if (err) {
                    this.logger.error(`Error fixing bug: ${err}`);
                    console.log("An error occurred while fixing the bug.");
                } else {
                    if (!bug) {
                        console.log("Bug ID not found.");
                    } else {
                        db.run("UPDATE bugs SET status = ? WHERE id = ?", ["Fixed", bugId], (err) => {
                            if (err) {
                                this.logger.error(`Error fixing bug: ${err}`);
                                console.log("An error occurred while fixing the bug.");
                            } else {
                                this.logger.info(`Bug ID ${bugId} has been marked as fixed.`);
                                console.log(`Bug ID ${bugId} has been marked as fixed.`);
                            }
                        });
                    }
                }
            });
            db.close();
        } catch (e) {
            this.logger.error(`Error fixing bug: ${e}`);
            console.log("An error occurred while fixing the bug.");
        }
    }
}

// Example usage:
const bugTracker = new BugTracker('robjam1990/Psychosis/Middleware/');

// Report a bug
bugTracker.reportBug("Encountered a glitch when accessing inventory.");

// View all reported bugs
console.log("All reported bugs:");
bugTracker.viewBugs();

// Assuming you know the bug ID from the viewBugs output, let's say it's "123e4567-e89b-12d3-a456-426614174000"
const bugIdToFix = "123e4567-e89b-12d3-a456-426614174000";
bugTracker.fixBug(bugIdToFix);

File: Middleware/BugTracking.py
# File = robjam1990/Psychosis/Middleware/BugTracking.py
import os
import uuid
import sqlite3
import logging

class BugTracker:
    def __init__(self, save_location):
        self.save_location = save_location
        self.bug_reports_file = os.path.join(save_location, 'bug_reports.db')
        self._initialize_database()
        self.logger = self._setup_logger()

    def _initialize_database(self):
        try:
            # Initialize database and table if not exists
            conn = sqlite3.connect(self.bug_reports_file)
            c = conn.cursor()
            c.execute('''CREATE TABLE IF NOT EXISTS bugs
                         (id TEXT PRIMARY KEY, description TEXT, status TEXT)''')
            conn.commit()
            conn.close()
        except Exception as e:
            self.logger.error(f"Error initializing database: {str(e)}")

    def _setup_logger(self):
        logger = logging.getLogger(__name__)
        logger.setLevel(logging.DEBUG)
        formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
        file_handler = logging.FileHandler(os.path.join(self.save_location, 'bug_tracker.log'))
        file_handler.setFormatter(formatter)
        logger.addHandler(file_handler)
        return logger

    def report_bug(self, bug_description):
        try:
            if not bug_description:
                raise ValueError("Bug description cannot be empty.")
            # Store bug reports in the database
            bug_id = str(uuid.uuid4())
            conn = sqlite3.connect(self.bug_reports_file)
            c = conn.cursor()
            c.execute("INSERT INTO bugs (id, description, status) VALUES (?, ?, ?)", (bug_id, bug_description, "Open"))
            conn.commit()
            conn.close()
            self.logger.info(f"Bug reported successfully. Bug ID: {bug_id}")
            print(f"Bug reported successfully. Bug ID: {bug_id}")
        except Exception as e:
            self.logger.error(f"Error reporting bug: {str(e)}")
            print("An error occurred while reporting the bug.")

    def view_bugs(self):
        try:
            # View all reported bugs
            conn = sqlite3.connect(self.bug_reports_file)
            c = conn.cursor()
            c.execute("SELECT * FROM bugs")
            bugs = c.fetchall()
            conn.close()
            if bugs:
                for bug in bugs:
                    print(f"{bug[0]} - {bug[1]} ({bug[2]})")
            else:
                print("No bugs reported.")
        except Exception as e:
            self.logger.error(f"Error viewing bugs: {str(e)}")
            print("An error occurred while viewing bugs.")

    def fix_bug(self, bug_id):
        try:
            conn = sqlite3.connect(self.bug_reports_file)
            c = conn.cursor()
            c.execute("SELECT * FROM bugs WHERE id=?", (bug_id,))
            bug = c.fetchone()
            if not bug:
                print("Bug ID not found.")
                return
            c.execute("UPDATE bugs SET status = ? WHERE id = ?", ("Fixed", bug_id))
            conn.commit()
            conn.close()
            self.logger.info(f"Bug ID {bug_id} has been marked as fixed.")
            print(f"Bug ID {bug_id} has been marked as fixed.")
        except Exception as e:
            self.logger.error(f"Error fixing bug: {str(e)}")
            print("An error occurred while fixing the bug.")

if __name__ == "__main__":
    # Example usage:
    bug_tracker = BugTracker(save_location='robjam1990/Psychosis/Middleware/')

    # Report a bug
    bug_tracker.report_bug("Encountered a glitch when accessing inventory.")

    # View all reported bugs
    print("All reported bugs:")
    bug_tracker.view_bugs()

    # Assuming you know the bug ID from the view_bugs output, let's say it's "123e4567-e89b-12d3-a456-426614174000"
    bug_id_to_fix = "123e4567-e89b-12d3-a456-426614174000"
    bug_tracker.fix_bug(bug_id_to_fix)
